src/progress.rs

```diff
@@ -348,6 +349,84 @@ impl ProgressState {
             (1_000_000_000 / avg_time) as u64
         }
     }
+
+    /// Call the provied `FnOnce` to update the state.  Then redraw the
```

Comment:
Would you mind isolating the pure refactoring (moving the finish updating code into the state into a separate commit or PR? That way I can more carefully review the other changes you're making here.